kernel/USER_GUIDE.md

```diff
@@ -0,0 +1,604 @@
+
+
+<table>
+  <tr>
+   <td><strong>Step</strong>
+   </td>
+   <td><strong>Typical query phase when this step occurs</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>Resolve the table snapshot to query
+   </td>
+   <td>Logical plan analysis phase when the plan's schema and other details need to resolved and validated
+   </td>
+  </tr>
+  <tr>
+   <td>Resolve files to scan based on query parameters
+   </td>
+   <td>Physical plan generation, when the final parameters of the scan are available. For example: 
+<ul>
+
+<li>Schema of data to read after pruning away unused columns
+
+<li>Query filters to apply after filter rearrangement
+</li>
+</ul>
+   </td>
+  </tr>
+  <tr>
+   <td>Distribute the file information to workers
+   </td>
+   <td>Physical plan execution, only if it is a distributed engine.
+   </td>
+  </tr>
+  <tr>
+   <td>Read the columnar data using the file information
+   </td>
+   <td>Physical plan execution, when the data is being processed by the engine 
+   </td>
+  </tr>
+</table>
+
+
+Let's understand the details of each step.
+
+
+#### Step 3.1: Resolve the table snapshot to query
+
+The first step is to resolve the consistent snapshot and the schema associated with it. This is often required by the connector / engine to resolve and validate the logical plan of the scan query (if the concept of logical plan exists in your engine). To achieve this, the connector has to do the following.
+
+
+
+* Resolve the table path from the query: If the path is directly available, then this is easy. Else if it is a query based on a catalog table (for example, a Delta table defined in Hive Metastore), then the connector has to resolve the table location / path from the catalog.
+* Initialize the TableClient object: Create a new instance of the TableClient that you have chosen in [Step 2](#bookmark=id.26in1rg). 
+* Initialize the Kernel objects and get the schema: Assuming the query is on the latest available version/snapshot of the table, you can get the table schema as follows:
+
+
+```
+import io.delta.kernel.*;
+import io.delta.kernel.defaults.*';
+
+TableClient myTableClient = new MyTableClient();
+Table myTable = Table.forPath(myTablePath);
+Snapshot mySnapshot = myTable.getLatestSnapshot(myTableClient);
+StructType mySchema = mySnapshot.getSchema(myTableClient);
+```
+
+
+If you want to query a specific version of the table (that is, not the schema), then you can get the required snapshot as ``myTable.getSnapshot(version)``.
```

Comment:
I don't quite understand this sentence? If the table's schema can change over time, it seems like we should always get the table's schema as-of whatever specific version we want to query?